### PR TITLE
Improve p2p message deserialization

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

File: neo/Network/P2P/Message.cs [hunk 1]
```diff
@@ -51,51 +52,11 @@ private void DecompressPayload()
...
-                    Payload = decompressed.AsSerializable<FilterLoadPayload>();
-                    break;
-                case MessageCommand.FilterAdd:
-                    Payload = decompressed.AsSerializable<FilterAddPayload>();
-                    break;
-                case MessageCommand.MerkleBlock:
-                    Payload = decompressed.AsSerializable<MerkleBlockPayload>();
-                    break;
-            }
+            MemberInfo[] members = typeof(MessageCommand).GetMember(Command.ToString());
```
Comment (shargon) [inline neo/Network/P2P/Message.cs]: We should pre-cache this in a `Dictionay<MessageCommand,Membernfo>`

File: neo/IO/Caching/ReflectionCache.cs [hunk 2]
```diff
@@ -1,77 +1,34 @@
...
-            return def;
-        }
-        /// <summary>
-        /// Create object from key
-        /// </summary>
-        /// <typeparam name="K">Type</typeparam>
-        /// <param name="key">Key</param>
-        /// <param name="def">Default value</param>
-        public K CreateInstance<K>(T key, K def = default(K))
+        public static object CreateInstance(T key, object def = null)
```
Comment (shargon) [inline neo/IO/Caching/ReflectionCache.cs]:
```suggestion
        public static TValue CreateInstance(T key, object def = null)
```

File: neo/IO/Caching/ReflectionCache.cs [hunk 3]
```diff
@@ -1,77 +1,34 @@
 using System;
 using System.Collections.Generic;
-using System.Linq;
 using System.Reflection;
 
 namespace Neo.IO.Caching
 {
-    public class ReflectionCache<T> : Dictionary<T, Type>
+    internal static class ReflectionCache<T> where T : Enum
```
Comment (shargon) [inline neo/IO/Caching/ReflectionCache.cs]:
```suggestion
    internal static class ReflectionCache<T, TValue> where T : Enum
```

File: neo/IO/Caching/ReflectionCache.cs (hunk 2 above)
Comment (erikzhang) [reply neo/IO/Caching/ReflectionCache.cs]: This method is not used now.

File: neo/IO/Caching/ReflectionCache.cs (hunk 3 above)
Comment (erikzhang) [reply neo/IO/Caching/ReflectionCache.cs]: What is this `TValue` for?

**Comment by erikzhang** [comment]
@lock9 This will not improve performance. But it ensures that we won't make mistakes when we add new p2p commands in the future. And improved readability.

**Pull request merged**